go/timer/timer.go

```diff
@@ -69,7 +69,7 @@ type Timer struct {
 // NewTimer creates a new Timer object
 func NewTimer(interval time.Duration) *Timer {
 	tm := &Timer{
-		msg: make(chan typeAction),
+		msg: make(chan typeAction, 1),
```

Comment:
Golang n00b question here... 

OK, so making it a buffered channel, even with a buffer size of 1, prevents the write to the channel from blocking in this case. I'm curious what rules we can apply here... 

A buffer size of one allows us to buffer 1 unread message in the channel: https://go.dev/play/p/CGi0hp-kbEs

And a buffer size of two allows us to buffer two: https://go.dev/play/p/MC4lKPxDrCk

It's not clear to me why that is, or is not, the right thing to do in this case. The Timer is generally intended to support exactly 1-2 multiple goroutines write to it concurrently w/o requiring the reads/writes be synchronized (using the mutex Timer has)?

Reply:
@mattlord Your concern is absolutely legitimate! I was thinking the same thing when I had thought of this solution. I had thought of 2 things - 1. Is it okay to increase the buffer size of the channel from 0 which will unblock some writes to it? and 2. How much should we increase it to? 1, 2 or 1000?

Let me answer both these questions and my thoughts as to how I arrived there.
To effectively answer these question lets look at the usages of `msg`, the channel in question.
There are 3 writers to the channel - `Stop`, `Trigger` and `SetInterval`. There is only 1 reader which is the `run` function. Let's look at each in detail -
- The purpose of the `Trigger` function to trigger a new run. Earlier it would return only after the run had started (but not finished). If the writes are buffered, then it will return after scheduling a run, but not waiting for it to start. From the outside the two situations are indistinguishable since the run function can take while to finish execution so in both cases, running a `Trigger` only guarantees you that the function will run eventually.
- The purpose of the `SetInterval` function is to change the timer intervals between each execution. Earlier, this function would return after the interval had been changed. Now the function returns after the interval change has been scheduled and after the current run (if any) finishes, that will take effect. The next run would still be with the new interval. Again there is no distinguishable change from the caller's perspective.
- The final function is `Stop`, which is used to stop the timer's execution. As per the comment of that function `It guarantees that the timer will not execute any more calls to keephouse once it has returned.` But there is no promise for the function which is executing now! If there is keep house function running, then in the run function we are blocked on the `msg` channel read, and we would immediately pick up the message stop pushed and not run any more keephouse functions. If we are executing the keephouse function, then it gets a little more interesting. According to the guarantee, it is okay to finish execution of this keephouse function, but we shouldn't start another one. So, if we buffer the message that Stop function is sending, we have to be sure that it gets processed after the current function's execution and but before a new execution starts. This is what constraints us to the buffer size 1! If we had a buffer size 2, then there could already be a message in the msg channel. Stop adds one more and returns, but the first messaged is processed first (which could be a trigger message) and this starts a new keephouse run even though Stop has returned!

So, keeping in mind all the observable changes and the promises of the timer module, we now have the answer to our two initial questions - 
1. It is safe to increase the buffer size.
2. But the buffer size can only be atmost 1, because we need to guarantee that the Stop functions message is at the top of the msg channel when it returns! (There are other ways to guarantee this, by increasing the buffer size more and then waiting in the stop function until it is empty, but we don't need to go there since that is not required)

Because of this analysis, I think it is safe to make this simple, albeit a large change as you point out.
I don't expect there to be any issues if the usages of the timer are written considering it's guarantees, which are still held. The only function that could potentially create issues is the `Stop` function which (now) returns after scheduling a stop but a keephouse function might still be in progress. Earlier we would have waited until everything had stopped.
However, in this case, I would argue that the usage of the timer `Stop` is incorrect and not this change, because we still uphold the guarantees and we would fix that place (if any).

I have run this by @harshit-gangal too. Does my analysis make sense?
cc - @deepthi if she has any insights if this would break something

Reply:
I'll do that 👍